src/transaction.ts

```diff
@@ -216,6 +216,160 @@ export const sendTransaction = async function (
+    if (accounts > 4) throw Error(`Account reference limit of 4 exceeded in transaction ${i}`)
+
+    const assets = group[i].txn.appForeignAssets?.length || 0
+    const apps = group[i].txn.appForeignApps?.length || 0
+    const boxes = group[i].txn.boxes?.length || 0
+
+    if (accounts + assets + apps + boxes > 8) throw Error(`Resource reference limit of 8 exceeded in transaction ${i}`)
+  })
+
+  const findTxnBelowRefLimit = (
+    txns: algosdk.TransactionWithSigner[],
+    type: 'account' | 'assetHolding' | 'appLocal' | 'other' = 'other',
+  ) => {
+    const txnIndex = txns.findIndex((t) => {
+      const accounts = t.txn.appAccounts?.length || 0
+      if (type === 'account') return accounts < 4
+
+      const assets = t.txn.appForeignAssets?.length || 0
+      const apps = t.txn.appForeignApps?.length || 0
+      const boxes = t.txn.boxes?.length || 0
+
+      if (type === 'assetHolding' || type === 'appLocal') {
+        return accounts + assets + apps + boxes < 7 && accounts < 4
+      }
+
+      return accounts + assets + apps + boxes < 8
+    })
+
+    if (txnIndex === -1) {
+      throw Error('No more transactions below reference limit. Add another app call to the group.')
+    }
+
+    return txnIndex
+  }
+
+  const g = unnamedResourcesAccessed.group
+
+  if (g) {
+    // Do cross-reference resources first because they are the most restrictive in terms
+    // of which transactions can be used
+    g.appLocals?.forEach((a) => {
+      const txnIndex = findTxnBelowRefLimit(group, 'appLocal')
+      group[txnIndex].txn.appForeignApps?.push(Number(a.app))
+      group[txnIndex].txn.appAccounts?.push(algosdk.decodeAddress(a.account))
+    })
+
+    g.assetHoldings?.forEach((a) => {
+      const txnIndex = findTxnBelowRefLimit(group, 'assetHolding')
+      group[txnIndex].txn.appForeignAssets?.push(Number(a.asset))
+      group[txnIndex].txn.appAccounts?.push(algosdk.decodeAddress(a.account))
+    })
+
+    // Do accounts next because the account limit is 4
+    g.accounts?.forEach((a) => {
+      const txnIndex = findTxnBelowRefLimit(group, 'account')
+      group[txnIndex].txn.appAccounts?.push(algosdk.decodeAddress(a))
+    })
+
+    g.boxes?.forEach((b) => {
+      const txnIndex = findTxnBelowRefLimit(group)
+      group[txnIndex].txn.boxes?.push({ appIndex: Number(b.app), name: b.name })
+    })
+
+    g.assets?.forEach((a) => {
+      const txnIndex = findTxnBelowRefLimit(group)
+      group[txnIndex].txn.appForeignAssets?.push(Number(a))
+    })
+
+    g.apps?.forEach((a) => {
+      const txnIndex = findTxnBelowRefLimit(group)
+      group[txnIndex].txn.appForeignApps?.push(Number(a))
+    })
+
+    if (g.extraBoxRefs) {
+      for (let i = 0; i < g.extraBoxRefs; i += 1) {
+        const txnIndex = findTxnBelowRefLimit(group)
+        group[txnIndex].txn.boxes?.push({ appIndex: 0, name: new Uint8Array(0) })
+      }
+    }
+  }
+
+  const newAtc = new algosdk.AtomicTransactionComposer()
+
+  group.forEach((t) => {
```

Comment:
wouldn't using clone be a bit simpler in this case? Not sure if it auto cleans the group field for you though

Reply:
[AtomicTransactionComposer](https://algorand.github.io/js-algorand-sdk/classes/AtomicTransactionComposer.html#clone.clone-1)

Reply:
Didn't even know this was a thing... I'll leverage it in the next commit

Reply:
Actually I realized I can't use this because I am mutating the transactions given by `atc.buildGroup()`